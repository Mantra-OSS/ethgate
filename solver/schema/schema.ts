import type { AksharaAbstract } from "@/akshara/index.js";
import {
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  printSchema,
} from "graphql";

import type {
  Block,
  ChainHasBlock,
  ChainHasTransaction,
  NodeAbstract,
  Transaction,
} from "../data/index.js";
import type { Chain } from "../data/index.js";
import type { EthgateSolverDatabase } from "../database/database.js";

import { GraphQLRelayCursor } from "./helpers/relay.js";
import { SchemaNodeInterface } from "./index.js";
import { asdasd } from "./node.js";

export class SchemaContext {
  readonly node: AksharaAbstract;
  readonly db: EthgateSolverDatabase;

  constructor(db: EthgateSolverDatabase) {
    this.node = db.provider;
    this.db = db;
  }
}

export class EthgateSolverSchema extends GraphQLSchema {
  constructor() {
    const { schemaEdgeTypes, schemaNodeTypes } = asdasd();
    const query = new GraphQLObjectType<void, SchemaContext>({
      name: "Query",
      fields: () => ({
        node: {
          type: SchemaNodeInterface,
          args: {
            id: {
              type: new GraphQLNonNull(GraphQLID),
              description: "The ID of the node",
            },
          },
          description: "The node",
          resolve(parent, args: { id: NodeAbstract["id"] }, ctx) {
            return ctx.db.getNode(args.id);
          },
        },
        root: {
          type: new GraphQLNonNull(schemaNodeTypes.get("Chain")!),
          resolve(parent, args, ctx) {
            return ctx.db.readNode<Chain>("Chain:1");
          },
        },
      }),
    });

    const subscription = new GraphQLObjectType<void, SchemaContext>({
      name: "Subscription",
      fields: () => ({
        chain_blocks: {
          type: new GraphQLObjectType({
            name: "ChainHasBlocksSubscription",
            fields: {
              edges: {
                type: new GraphQLNonNull(
                  new GraphQLList(
                    new GraphQLNonNull(schemaEdgeTypes.get("ChainHasBlock")!)
                  )
                ),
                description: "The edges of the subscription",
              },
            },
            description: "The network blocks subscription",
          }),
          args: {
            id: {
              type: new GraphQLNonNull(GraphQLID),
              description: "The ID of an network object",
            },
            first: {
              type: GraphQLInt,
              description: "The number of blocks to return",
            },
            before: {
              type: GraphQLRelayCursor,
              description: "The cursor to start from",
            },
          },
          description: "The network blocks subscription",
          async *subscribe(
            parent,
            args: {
              id: Chain["id"];
              first?: number;
              before?: Block["id"];
            },
            ctx
          ): AsyncGenerator<{
            chain_blocks: { edges: ChainHasBlock[] };
          } | void> {
            const chain = await ctx.db.readNode<Chain>(args.id);
            const updates = ctx.db.networkUpdates(chain.id);
            let before = args.before;

            let updatePromise = updates.next();
            while (true) {
              if (!before) {
                const item = await ctx.db
                  .getConnection<ChainHasBlock>("ChainHasBlock", chain.id, {
                    first: 1,
                  })
                  .next();
                if (item.done) {
                  await updatePromise;
                  updatePromise = updates.next();
                  yield;
                  continue;
                }
                const edge = item.value;
                before = edge.headId;
              }

              const connection = ctx.db.getConnection<ChainHasBlock>(
                "ChainHasBlock",
                chain.id,
                {
                  first: args.first,
                  before,
                }
              );
              // TODO: Do not collect, but yield in batches
              const { edges } = await connection.collect();
              yield { chain_blocks: { edges: edges.reverse() } };

              await updatePromise;
              updatePromise = updates.next();
              yield;
            }
          },
        },
        chain_transactions: {
          type: new GraphQLObjectType({
            name: "ChainHasTransactionsSubscription",
            fields: {
              edges: {
                type: new GraphQLNonNull(
                  new GraphQLList(
                    new GraphQLNonNull(
                      schemaEdgeTypes.get("ChainHasTransaction")!
                    )
                  )
                ),
                description: "The edges of the subscription",
              },
            },
            description: "The network transactions subscription",
          }),
          args: {
            id: {
              type: new GraphQLNonNull(GraphQLID),
              description: "The ID of an network object",
            },
            first: {
              type: GraphQLInt,
              description: "The number of transactions to return",
            },
            before: {
              type: GraphQLRelayCursor,
              description: "The cursor to start from",
            },
          },
          description: "The network transactions subscription",
          async *subscribe(
            parent,
            args: {
              id: Chain["id"];
              first?: number;
              before?: Transaction["id"];
            },
            ctx
          ): AsyncGenerator<{
            chain_transactions: { edges: ChainHasTransaction[] };
          } | void> {
            const chain = await ctx.db.readNode<Chain>(args.id);
            const updates = ctx.db.networkUpdates(chain.id);
            let before = args.before;

            let updatePromise = updates.next();
            while (true) {
              if (!before) {
                const item = await ctx.db
                  .getConnection<ChainHasTransaction>(
                    "ChainHasTransaction",
                    chain.id,
                    {
                      first: 1,
                    }
                  )
                  .next();
                if (item.done) {
                  await updatePromise;
                  updatePromise = updates.next();
                  yield;
                  continue;
                }
                const edge = item.value;
                before = edge.headId;
              }

              const connection = ctx.db.getConnection<ChainHasTransaction>(
                "ChainHasTransaction",
                chain.id,
                {
                  first: args.first,
                  before,
                }
              );
              // TODO: Do not collect, but yield in batches
              const { edges } = await connection.collect();
              yield { chain_transactions: { edges: edges.reverse() } };

              await updatePromise;
              updatePromise = updates.next();
              yield;
            }
          },
        },
        greetings: {
          type: GraphQLString,
          subscribe: async function* () {
            for (const hi of ["Hi", "Bonjour", "Hola", "Ciao", "Zdravo"]) {
              yield { greetings: hi };
            }
          },
        },
      }),
    });

    super({
      query,
      subscription,
    });
  }

  print(): string {
    return printSchema(this);
  }
}
