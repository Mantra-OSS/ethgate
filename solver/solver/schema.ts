import type { ConnectionData } from '@/app/client/backend';
import { GraphQLInterfaceType } from 'graphql';
import {
  ASTNode,
  GraphQLBoolean,
  GraphQLError,
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLScalarTypeConfig,
  GraphQLSchema,
  GraphQLString,
  Kind,
  ObjectValueNode,
  Source,
  ValueNode,
  printSchema,
} from 'graphql';

import type { NodeType, SolverEdge, SolverNode } from '../graph';

import { GraphQLJSONObject } from './schema/scalars';
import type { Solver, SolverGraph } from './solver';

export type SolverSchema = GraphQLSchema;
export const createSolverSchema = (graph: SolverGraph): SolverSchema => {
  const pageInfo = new GraphQLObjectType({
    name: 'PageInfo',
    fields: {
      hasNextPage: {
        type: new GraphQLNonNull(GraphQLBoolean),
      },
      hasPreviousPage: {
        type: new GraphQLNonNull(GraphQLBoolean),
      },
      startCursor: {
        type: GraphQLString,
      },
      endCursor: {
        type: GraphQLString,
      },
    },
  });
  const nodeMeta = new GraphQLObjectType({
    name: 'NodeMeta',
    fields: {
      name: {
        type: new GraphQLNonNull(GraphQLString),
      },
      slug: {
        type: new GraphQLNonNull(GraphQLString),
      },
      chainId: {
        type: GraphQLID,
      },
    },
  });
  const nodeInterface: GraphQLInterfaceType = new GraphQLInterfaceType({
    name: 'Node',
    fields: () => ({
      id: {
        type: new GraphQLNonNull(GraphQLID),
      },
      meta: {
        type: new GraphQLNonNull(nodeMeta),
      },
      data: {
        type: new GraphQLNonNull(GraphQLJSONObject),
      },
      connection: {
        type: new GraphQLNonNull(connectionInterface),
        args: {
          type: {
            type: new GraphQLNonNull(GraphQLString),
          },
          first: {
            type: GraphQLInt,
          },
          after: {
            type: GraphQLString,
          },
        },
      },
    }),
    resolveType: (value: SolverNode) => value.type,
  });
  const edgeInterface = new GraphQLInterfaceType({
    name: 'Edge',
    fields: {
      type: {
        type: new GraphQLNonNull(GraphQLString),
      },
      tailId: {
        type: new GraphQLNonNull(GraphQLID),
      },
      tail: {
        type: new GraphQLNonNull(nodeInterface),
      },
      headId: {
        type: new GraphQLNonNull(GraphQLID),
      },
      head: {
        type: new GraphQLNonNull(nodeInterface),
      },
      node: {
        type: new GraphQLNonNull(nodeInterface),
      },
      cursor: {
        type: new GraphQLNonNull(GraphQLString),
      },
    },
    resolveType: (value: SolverEdge) => `${value.type}Edge`,
  });
  const connectionInterface = new GraphQLInterfaceType({
    name: 'Connection',
    fields: {
      edges: {
        type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(edgeInterface))),
      },
      pageInfo: {
        type: new GraphQLNonNull(pageInfo),
      },
    },
    resolveType: (value: ConnectionData) => `${value.type}Connection`,
  });
  const nodes: GraphQLObjectType[] = graph.nodeTypes.map((nodeType) => {
    const nodeData = GraphQLJSONObject;
    // const nodeData = new GraphQLObjectType({
    //   name: `${nodeType.name}Data`,
    //   fields: () => {
    //     return {
    //       ...Object.fromEntries(
    //         nodeType.
    //     }
    //   }
    // })
    return new GraphQLObjectType<SolverNode, Solver>({
      name: nodeType.name,
      interfaces: [nodeInterface],
      fields: () => {
        const edgeTypes = graph.getEdgeTypesForNode(nodeType.name);
        return {
          id: {
            type: new GraphQLNonNull(GraphQLID),
          },
          meta: {
            type: new GraphQLNonNull(nodeMeta),
          },
          data: {
            type: new GraphQLNonNull(nodeData),
          },
          connection: {
            type: new GraphQLNonNull(connectionInterface),
            args: {
              type: {
                type: new GraphQLNonNull(GraphQLString),
              },
              first: {
                type: GraphQLInt,
              },
              after: {
                type: GraphQLString,
              },
            },
            async resolve(parent, args, solver): Promise<ConnectionData> {
              return {
                ...(await solver.database.getConnection(args.type, parent.id, args).collect()),
                type: args.type,
                tailId: parent.id,
              };
            },
          },
          ...Object.fromEntries(
            edgeTypes.map((edgeType) => {
              const connection = connections.find(
                (connection) => connection.name === `${edgeType.typeName}Connection`,
              );
              if (!connection) throw new Error(`connection not found: ${edgeType.typeName}`);
              return [
                edgeType.connectionName,
                {
                  type: new GraphQLNonNull(connection),
                  args: {
                    first: {
                      type: GraphQLInt,
                    },
                    after: {
                      type: GraphQLString,
                    },
                  },
                },
              ];
            }),
          ),
        };
      },
    });
  });

  const edges = graph.edgeTypes.map((edgeType) => {
    return new GraphQLObjectType({
      name: `${edgeType.typeName}Edge`,
      interfaces: [edgeInterface],
      fields: () => {
        const tail = nodes.find((node) => node.name === edgeType.tail.name);
        if (!tail) throw new Error(`tail not found: ${edgeType.tail.name}`);
        const head = nodes.find((node) => node.name === edgeType.head.name);
        if (!head) throw new Error(`head not found: ${edgeType.head.name}`);
        return {
          type: {
            type: new GraphQLNonNull(GraphQLString),
          },
          tailId: {
            type: new GraphQLNonNull(GraphQLID),
          },
          tail: {
            type: new GraphQLNonNull(tail),
          },
          headId: {
            type: new GraphQLNonNull(GraphQLID),
          },
          head: {
            type: new GraphQLNonNull(head),
          },
          node: {
            type: new GraphQLNonNull(head),
          },
          cursor: {
            type: new GraphQLNonNull(GraphQLString),
          },
        };
      },
    });
  });

  const connections = graph.edgeTypes.map((edgeType) => {
    return new GraphQLObjectType({
      name: `${edgeType.typeName}Connection`,
      interfaces: [connectionInterface],
      fields: () => {
        const edge = edges.find((edge) => edge.name === `${edgeType.typeName}Edge`);
        if (!edge) throw new Error(`edge not found: ${edgeType.typeName}`);
        return {
          edges: {
            type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(edge))),
          },
          pageInfo: {
            type: new GraphQLNonNull(pageInfo),
          },
        };
      },
    });
  });

  const root = nodes.find((node) => node.name === 'Explorer');
  if (!root) throw new Error(`root not found`);

  const query = new GraphQLObjectType<undefined, Solver>({
    name: 'Query',
    fields: {
      node: {
        type: nodeInterface,
        args: {
          id: {
            type: new GraphQLNonNull(GraphQLID),
          },
        },
        resolve(_, { id }, solver) {
          return solver.database.getNode(id);
        },
      },
      root: {
        type: new GraphQLNonNull(root),
        resolve(parent, args, solver) {
          return solver.getRoot();
        },
      },
    },
  });

  const schema = new GraphQLSchema({
    query,
  });
  return schema;
};
