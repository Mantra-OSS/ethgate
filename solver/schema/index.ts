export * from './helpers/graph.js';
export * from './schema.js';
