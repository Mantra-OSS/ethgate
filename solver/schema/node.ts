import type { AksharaChainId, Hash } from "@/akshara/index.js";
import { GraphQLInt, GraphQLNonNull, GraphQLString } from "graphql";

import {
  BlockHasLog,
  BlockHasReceipt,
  BlockHasTransaction,
  Chain,
  ChainHasBlock,
  ChainHasChain,
  ChainHasDescendantBlock,
  ChainHasTransaction,
  Log,
  Receipt,
  ReceiptHasLog,
  Transaction,
} from "../data/index.js";
import { Block } from "../data/index.js";

import { SchemaEdgeType, SchemaNodeType } from "./helpers/graph.js";
import { GraphQLU64 } from "./scalars/primitive.js";

export const asdasd = () => {
  const edgeTypes = [
    ChainHasChain,
    ChainHasBlock,
    ChainHasDescendantBlock,
    ChainHasTransaction,
    BlockHasTransaction,
    BlockHasLog,
    BlockHasReceipt,
    ReceiptHasLog,
  ];

  const schemaNodeTypes = new Map<string, SchemaNodeType<any>>();

  const schemaEdgeTypes = new Map<string, SchemaEdgeType<any, any>>();

  const SchemaChain: SchemaNodeType<Chain> = new SchemaNodeType(
    schemaNodeTypes,
    schemaEdgeTypes,
    Chain.type as any,
    {
      name: "Chain",
      description: "A blockchain",
      fields: () => ({
        chain: {
          type: schemaNodeTypes.get("Chain")!,
          args: {
            chainId: {
              type: new GraphQLNonNull(GraphQLString),
              description: "The EIP155 ID of the chain",
            },
          },
          resolve(parent, args: { chainId: AksharaChainId }, ctx) {
            return ctx.node
              .getObject({ type: "Chain", chainId: args.chainId })
              .then((data) => data && new Chain(data));
          },
        },
        block: {
          type: SchemaBlock,
          args: {
            hash: {
              type: GraphQLString,
            },
            number: {
              type: GraphQLU64,
            },
          },
          resolve: (
            node,
            args: {
              hash?: Hash;
              number?: number;
            },
            ctx
          ) => {
            if (args.hash) {
              return ctx.node
                .getObject({
                  type: "Block",
                  chainId: node.data.chainId,
                  hash: args.hash,
                })
                .then((data) => data && new Block(data));
            } else if (typeof args.number === "number") {
              return ctx.node
                .getObject({
                  type: "Block",
                  chainId: node.data.chainId,
                  number: args.number,
                })
                .then((data) => data && new Block(data));
            }
          },
        },
        transaction: {
          type: SchemaTransaction,
          args: {
            hash: {
              type: GraphQLString,
            },
            blockNumber: {
              type: GraphQLInt,
            },
            transactionIndex: {
              type: GraphQLInt,
            },
          },
          resolve: (
            node,
            args: {
              hash?: Hash;
              blockNumber?: number;
              transactionIndex?: number;
            },
            ctx
          ) => {
            if (args.hash) {
              return ctx.node
                .getObject({
                  type: "Transaction",
                  chainId: node.data.chainId,
                  hash: args.hash,
                })
                .then((data) => data && new Transaction(data));
            } else if (args.blockNumber && args.transactionIndex) {
              return ctx.node
                .getObject({
                  type: "Transaction",
                  chainId: node.data.chainId,
                  blockNumber: args.blockNumber,
                  transactionIndex: args.transactionIndex,
                })
                .then((data) => data && new Transaction(data));
            }
          },
        },
        // Node fields
        parent: {
          type: SchemaChain,
          description: "The parent chain",
          resolve: (node, args, ctx) =>
            node.parentId && ctx.db.readNode<Chain>(node.parentId),
        },
      }),
    }
  );

  const SchemaBlock: SchemaNodeType<Block> = new SchemaNodeType(
    schemaNodeTypes,
    schemaEdgeTypes,
    Block,
    {
      name: "Block",
      description: "The block object",
      fields: () => ({
        transaction: {
          type: SchemaTransaction,
          args: {
            hash: {
              type: GraphQLString,
            },
            transactionIndex: {
              type: GraphQLInt,
            },
          },
          resolve: (
            node,
            args: {
              hash?: Hash;
              transactionIndex?: number;
            },
            ctx
          ) => {
            const chainId = node.data.chainId;
            if (args.hash) {
              return ctx.node
                .getTransaction({ chainId, hash: args.hash })
                .then((data) => data && new Transaction(data));
            } else if (typeof args.transactionIndex === "number") {
              const hash = node.data.transactions[args.transactionIndex];
              return ctx.node
                .getTransaction({ chainId, hash })
                .then((data) => data && new Transaction(data));
            }
          },
        },
        chain: {
          type: new GraphQLNonNull(SchemaChain),
          description: "The chain",
          resolve: (node, args, ctx) => ctx.db.readNode<Chain>(node.chainId),
        },
        // Node fields
        parent: {
          type: new GraphQLNonNull(SchemaBlock),
          resolve: (node, args, ctx) => ctx.db.readNode<Block>(node.parentId),
        },
      }),
    }
  );

  const SchemaTransaction: SchemaNodeType<Transaction> = new SchemaNodeType(
    schemaNodeTypes,
    schemaEdgeTypes,
    Transaction,
    {
      name: "Transaction",
      fields: () => ({
        // Node fields
        receipt: {
          type: new GraphQLNonNull(SchemaReceipt),
          resolve: (node, args, ctx) =>
            ctx.db.readNode<Receipt>(node.receiptId),
        },
        chain: {
          type: new GraphQLNonNull(SchemaChain),
          description: "The chain",
          resolve: (node, args, ctx) => ctx.db.readNode<Chain>(node.chainId),
        },
        block: {
          type: new GraphQLNonNull(SchemaBlock),
          resolve: (node, args, ctx) => ctx.db.readNode<Block>(node.blockId),
        },
      }),
    }
  );

  const SchemaReceipt: SchemaNodeType<Receipt> = new SchemaNodeType(
    schemaNodeTypes,
    schemaEdgeTypes,
    Receipt,
    {
      name: "Receipt",
      fields: () => ({
        log: {
          type: SchemaLog,
          args: {
            logIndex: {
              type: GraphQLInt,
            },
          },
          resolve: (node, args, ctx) => {
            return ctx.node
              .getLog({
                chainId: node.data.chainId,
                transactionHash: node.data.transactionHash,
                logIndex: args.logIndex,
              })
              .then((data) => data && new Log(data));
          },
        },
        // Node fields
        transaction: {
          type: new GraphQLNonNull(SchemaTransaction),
          resolve: (node, args, ctx) =>
            ctx.db.readNode<Transaction>(node.transactionId),
        },
        chain: {
          type: new GraphQLNonNull(SchemaChain),
          description: "The chain",
          resolve: (node, args, ctx) => ctx.db.readNode<Chain>(node.chainId),
        },
        block: {
          type: new GraphQLNonNull(SchemaBlock),
          resolve: (node, args, ctx) => ctx.db.readNode<Block>(node.blockId),
        },
      }),
    }
  );

  const SchemaLog: SchemaNodeType<Log> = new SchemaNodeType(
    schemaNodeTypes,
    schemaEdgeTypes,
    Log,
    {
      name: "Log",
      fields: () => ({
        // Node fields
        receipt: {
          type: new GraphQLNonNull(SchemaReceipt),
          resolve: (node, args, ctx) =>
            ctx.db.readNode<Receipt>(node.receiptId),
        },
        transaction: {
          type: new GraphQLNonNull(SchemaTransaction),
          resolve: (node, args, ctx) =>
            ctx.db.readNode<Transaction>(node.transactionId),
        },
        chain: {
          type: new GraphQLNonNull(SchemaChain),
          description: "The chain",
          resolve: (node, args, ctx) => ctx.db.readNode<Chain>(node.chainId),
        },
        block: {
          type: new GraphQLNonNull(SchemaBlock),
          resolve: (node, args, ctx) => ctx.db.readNode<Block>(node.blockId),
        },
      }),
    }
  );

  schemaNodeTypes.set("Chain", SchemaChain);
  schemaNodeTypes.set("Block", SchemaBlock);
  schemaNodeTypes.set("Transaction", SchemaTransaction);
  schemaNodeTypes.set("Receipt", SchemaReceipt);
  schemaNodeTypes.set("Log", SchemaLog);

  edgeTypes.forEach((edgeType) => {
    const asd = new SchemaEdgeType(schemaNodeTypes, edgeType, {
      name: edgeType.name,
    });
    schemaEdgeTypes.set(edgeType.name, asd);
  });

  return {
    schemaNodeTypes,
    schemaEdgeTypes,
  };
};
