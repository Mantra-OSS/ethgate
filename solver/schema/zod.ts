import type { GraphQLOutputType } from 'graphql';
import { GraphQLBoolean, GraphQLList, GraphQLObjectType, GraphQLString } from 'graphql';

import {
  GraphQLAddress,
  GraphQLChainId,
  GraphQLHash,
  GraphQLHex,
  GraphQLU64,
} from './scalars/primitive.js';

export function schemaToGql(schema: any): GraphQLOutputType | void {
  switch (schema.type) {
    case 'object': {
      if (schema.aksharaType) {
        return new GraphQLObjectType({
          name: schema.aksharaType,
          fields: {},
        });
      }
      return;
    }
    case 'string': {
      switch (schema.aksharaType) {
        case 'Hex': {
          return GraphQLHex;
        }
        case 'Hash': {
          return GraphQLHash;
        }
        case 'Address': {
          return GraphQLAddress;
        }
        case 'ChainId': {
          return GraphQLChainId;
        }
      }
      return GraphQLString;
    }
    case 'number': {
      return GraphQLU64;
    }
    case 'boolean': {
      return GraphQLBoolean;
    }
    case 'array': {
      const elem = schemaToGql(schema.items);
      if (!elem) {
        return;
      }
      return new GraphQLList(elem);
    }
    default: {
      throw new Error(`Unknown type ${schema.type}`);
    }
  }
}
