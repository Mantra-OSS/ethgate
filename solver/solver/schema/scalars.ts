import type { ASTNode, GraphQLScalarTypeConfig, ObjectValueNode, Source, ValueNode } from 'graphql';
import { GraphQLError, GraphQLScalarType, Kind, versionInfo } from 'graphql';

export const GraphQLJSONObjectConfig = {
  name: 'JSONObject',
  serialize: ensureObject,
  parseValue: ensureObject,
  parseLiteral: parseObject,
  extensions: {
    codegenScalarType: 'Record<string, any>',
    jsonSchema: {
      type: 'object',
      additionalProperties: true,
    },
  },
} as GraphQLScalarTypeConfig<object, object>;

export const GraphQLJSONObject = new GraphQLScalarType(GraphQLJSONObjectConfig);

export function identity<T>(value: T): T {
  return value;
}

export function ensureObject(value: any, ast?: ValueNode): object {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw createGraphQLError(
      `JSONObject cannot represent non-object value: ${value}`,
      ast
        ? {
            nodes: ast,
          }
        : undefined,
    );
  }

  return value;
}

export function parseObject(ast: ObjectValueNode, variables: any): any {
  const value = Object.create(null);
  ast.fields.forEach((field) => {
    value[field.name.value] = parseLiteral(field.value, variables);
  });

  return value;
}

export function parseLiteral(ast: ValueNode, variables: any): any {
  switch (ast.kind) {
    case Kind.STRING:
    case Kind.BOOLEAN:
      return ast.value;
    case Kind.INT:
    case Kind.FLOAT:
      return parseFloat(ast.value);
    case Kind.OBJECT:
      return parseObject(ast, variables);
    case Kind.LIST:
      return ast.values.map((n) => parseLiteral(n, variables));
    case Kind.NULL:
      return null;
    case Kind.VARIABLE: {
      const name = ast.name.value;
      return variables ? variables[name] : undefined;
    }
  }
}

interface GraphQLErrorOptions {
  nodes?: ReadonlyArray<ASTNode> | ASTNode | null;
  source?: Source;
  positions?: ReadonlyArray<number>;
  path?: ReadonlyArray<string | number>;
  originalError?: Error & {
    readonly extensions?: unknown;
  };
  extensions?: any;
}

export function createGraphQLError(message: string, options?: GraphQLErrorOptions): GraphQLError {
  if (versionInfo.major >= 17) {
    return new GraphQLError(message, options);
  }
  return new GraphQLError(
    message,
    options?.nodes,
    options?.source,
    options?.positions,
    options?.path,
    options?.originalError,
    options?.extensions,
  );
}
