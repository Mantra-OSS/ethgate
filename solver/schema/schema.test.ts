import { describe, expect, it } from '@jest/globals';
import { writeFile } from 'node:fs/promises';
import { resolve } from 'node:path';

import { EthgateSolverSchema } from './schema.js';

describe('PunkerSchema', () => {
  it.skip('has the same print', async () => {
    const schema = new EthgateSolverSchema();
    const printed = schema.print();

    expect(printed).toMatchSnapshot();
  });
  // A hack to dump the latest schema to data/schema.graphql
  it('can dump schema', async () => {
    const schema = new EthgateSolverSchema();
    const printed = schema.print();

    const platformPath = process.platform === 'win32' ? 'file:///' : 'file://';
    const path = resolve(
      import.meta.url.replace(platformPath, ''),
      '../../../../../apps/explorer/data/schema.graphql',
    );
    await writeFile(path, printed + '\n');
  });
});
