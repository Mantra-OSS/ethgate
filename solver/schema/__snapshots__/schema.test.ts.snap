// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PunkerSchema has the same print 1`] = `
"type Query {
  """The node"""
  node(
    """The ID of the node"""
    id: ID!
  ): Node
  root: Chain!
}

"""The node interface"""
interface Node {
  """The ID of the node"""
  id: ID!

  """The meta data of the node"""
  meta: NodeMeta!
}

"""The meta data of a node"""
type NodeMeta {
  """The type of the node"""
  type: String!

  """The local ID of the node"""
  localId: String!

  """The data of the node"""
  data: JSON!

  """The time of the node"""
  time: String!

  """The name of the node"""
  name: String!
}

"""
The \`JSON\` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSON

"""A blockchain"""
type Chain implements Node {
  id: ID!
  meta: NodeMeta!
  chainId: String!
  name: String!
  parentId: String
  chains(
    """The number of nodes to return after the cursor"""
    first: Int

    """The cursor after which to return nodes"""
    after: Cursor

    """The number of nodes to return before the cursor"""
    last: Int

    """The cursor before which to return nodes"""
    before: Cursor
  ): ChainHasChainConnection!
  blocks(
    """The number of nodes to return after the cursor"""
    first: Int

    """The cursor after which to return nodes"""
    after: Cursor

    """The number of nodes to return before the cursor"""
    last: Int

    """The cursor before which to return nodes"""
    before: Cursor
  ): ChainHasBlockConnection!
  descendantBlocks(
    """The number of nodes to return after the cursor"""
    first: Int

    """The cursor after which to return nodes"""
    after: Cursor

    """The number of nodes to return before the cursor"""
    last: Int

    """The cursor before which to return nodes"""
    before: Cursor
  ): ChainHasDescendantBlockConnection!
  transactions(
    """The number of nodes to return after the cursor"""
    first: Int

    """The cursor after which to return nodes"""
    after: Cursor

    """The number of nodes to return before the cursor"""
    last: Int

    """The cursor before which to return nodes"""
    before: Cursor
  ): ChainHasTransactionConnection!
  chain(
    """The EIP155 ID of the chain"""
    id: String!
  ): Chain
  block(hash: String, number: U64): Block
  transaction(hash: String): Transaction

  """The parent chain"""
  parent: Chain
}

type ChainHasChainConnection implements Connection {
  edges: [ChainHasChainEdge!]!
  pageInfo: PageInfo!
}

"""A connection to a list of items"""
interface Connection {
  """A list of edges"""
  edges: [Edge]

  """Information to aid in pagination"""
  pageInfo: PageInfo!
}

"""An edge in a connection"""
interface Edge {
  """The item at the end of the edge"""
  node: Node

  """A cursor for use in pagination"""
  cursor: String!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: Cursor

  """When paginating forwards, the cursor to continue."""
  endCursor: Cursor
}

"""A cursor for use in pagination"""
scalar Cursor

type ChainHasChainEdge implements Edge {
  node: Chain!
  cursor: String!
}

type ChainHasBlockConnection implements Connection {
  edges: [ChainHasBlockEdge!]!
  pageInfo: PageInfo!
}

type ChainHasBlockEdge implements Edge {
  node: Block!
  cursor: String!
}

"""The block object"""
type Block implements Node {
  id: ID!
  meta: NodeMeta!
  chainId: String!
  timestamp: U64!
  hash: Hash!
  number: U64!
  parentHash: Hash!
  gasLimit: U64!
  gasUsed: U64!
  baseFeePerGas: U64!
  logsBloom: Hex!
  miner: Address!
  size: U64!
  transactions(
    """The number of nodes to return after the cursor"""
    first: Int

    """The cursor after which to return nodes"""
    after: Cursor

    """The number of nodes to return before the cursor"""
    last: Int

    """The cursor before which to return nodes"""
    before: Cursor
  ): BlockHasTransactionConnection!
  receipts(
    """The number of nodes to return after the cursor"""
    first: Int

    """The cursor after which to return nodes"""
    after: Cursor

    """The number of nodes to return before the cursor"""
    last: Int

    """The cursor before which to return nodes"""
    before: Cursor
  ): BlockHasReceiptConnection!
  transaction(hash: String, index: Int): Transaction

  """The chain"""
  chain: Chain!
  parent: Block!
  transactionCount: Int!
}

scalar U64

scalar Hash

scalar Hex

scalar Address

type BlockHasTransactionConnection implements Connection {
  edges: [BlockHasTransactionEdge!]!
  pageInfo: PageInfo!
}

type BlockHasTransactionEdge implements Edge {
  node: Transaction!
  cursor: String!
}

type Transaction implements Node {
  id: ID!
  meta: NodeMeta!
  chainId: String!
  blockHash: Hash!
  blockNumber: U64!
  from: Address!
  gas: U64!
  gasPrice: U64!
  hash: Hash!
  input: Hex!
  nonce: Hex!
  to: Address!
  transactionIndex: U64!
  value: Hex!
  v: Hex!
  r: Hex!
  s: Hex!
  receipt: Receipt!

  """The chain"""
  chain: Chain!
  block: Block!
}

type Receipt implements Node {
  id: ID!
  meta: NodeMeta!
  chainId: String!
  blockHash: Hash!
  blockNumber: U64!
  contractAddress: String
  cumulativeGasUsed: U64!
  from: Address!
  gasUsed: U64!
  logsBloom: Hex!
  status: Hex!
  to: Address!
  transactionHash: Hash!
  transactionIndex: U64!
  logs(
    """The number of nodes to return after the cursor"""
    first: Int

    """The cursor after which to return nodes"""
    after: Cursor

    """The number of nodes to return before the cursor"""
    last: Int

    """The cursor before which to return nodes"""
    before: Cursor
  ): ReceiptHasLogConnection!
  log(index: Int): Log
  block: Block!
}

type ReceiptHasLogConnection implements Connection {
  edges: [ReceiptHasLogEdge!]!
  pageInfo: PageInfo!
}

type ReceiptHasLogEdge implements Edge {
  node: Log!
  cursor: String!
}

type Log implements Node {
  id: ID!
  meta: NodeMeta!
  chainId: String!
  address: Address!
  topics: [Hash]!
  data: Hex!
  blockNumber: U64!
  transactionHash: Hash!
  transactionIndex: U64!
  blockHash: Hash!
  logIndex: U64!
  removed: Boolean!
  block: Block!
  transaction: Block!
  receipt: Block!
}

type BlockHasReceiptConnection implements Connection {
  edges: [BlockHasReceiptEdge!]!
  pageInfo: PageInfo!
}

type BlockHasReceiptEdge implements Edge {
  node: Receipt!
  cursor: String!
}

type ChainHasDescendantBlockConnection implements Connection {
  edges: [ChainHasDescendantBlockEdge!]!
  pageInfo: PageInfo!
}

type ChainHasDescendantBlockEdge implements Edge {
  node: Block!
  cursor: String!
}

type ChainHasTransactionConnection implements Connection {
  edges: [ChainHasTransactionEdge!]!
  pageInfo: PageInfo!
}

type ChainHasTransactionEdge implements Edge {
  node: Transaction!
  cursor: String!
}

type Subscription {
  """The network blocks subscription"""
  chain_blocks(
    """The ID of an network object"""
    id: ID!

    """The number of blocks to return"""
    first: Int

    """The cursor to start from"""
    before: Cursor
  ): ChainHasBlocksSubscription
  greetings: String
}

"""The network blocks subscription"""
type ChainHasBlocksSubscription {
  """The edges of the subscription"""
  edges: [ChainHasBlockEdge!]!
}"
`;
